### DISCUSSION TITLE
Implement Mathematics Rendering

### DISCUSSION THREAD

**Discussion by RyanGreenup** [GitJournal/GitJournal#815] [Feature Requests]
Writing Math is crucial for me and I would love it if git journal could support this.

[Mathjax](https://www.mathjax.org/) would be the most ideal because it has such a large range of functions but [KaTeX](https://katex.org/) may be easier to implement and makes a little more sense in the context of markdown.

**Comment by vHanda** [comment]
I did a small proof of concept yesterday with Katex. I managed to render it. I'll try to get it done within the week.

I don't know almost anything about latex, I simply chose Katex cause it's supposed to be faster. Let me start with Katex, then maybe I can even add MathJax.

**Comment by RyanGreenup** [comment]
@vHanda I'm so glad to hear that you're implementing this! I've needed something like this app for so long.

The basic distinction that affects users between KaTeX, LaTeX and MathJax is related to aligning equations, KaTeX is more than good enough and in some ways makes more sense. I'll list the main differences below in case it's of any help to anybody (I know I was confused for a while).

### LaTeX and MathJax
In LaTeX and MathJax the function `\tag{3.3}` will put a (3.3) next to the equation and  `\label{name}` let's you refer to an equation later on by using `$\eqref{name}` but the catch is that it has to be inside a `\begin{align} ... \end{align}` environment, this environment cannot be delimited with `$$ ... $$` or LaTeX won't compile (but it won't render without `$$` in MathJax so doing this you lose the luxury of being able to `pandoc` between).

It is also possible to use `$$ \begin{aligned} ... \end{aligned} $$`, but this will only support displaying those numbers if you dispense with `$$` and instead replace them with equation tags like this `\begin{equation} \begin{aligned} ... \end{aligned} \end{equation}`, again though this won't render in MathJax.

The next issue is that MathJax won't render anything that isn't between `$$` so it can be a headache to number equations.

### KaTeX
In KaTeX the `align` environment isn't supported, instead you have to use `$$\begin{aligned}...\end{aligned}$$`, this is in a sense good because it prevents the erroneous nesting, so converting is easier.

However to get around the lack of support for numbering equations they implemented the `\tag{8.7}` function (e.g. this would display (8.7)), again this is confusing because this will stop LaTeX from compiling, so it's another thing to bear in mind.

### Cross Platform Manual Numbering
So my basic solution is to number manually using:

```
$$\begin{aligned}
4x &= 4 \qquad      \ldots \left( 8.3 \right)
$$\begin{aligned}
```

**Comment by vHanda** [comment]
![photo_2020-05-19 18 39 45](https://user-images.githubusercontent.com/426467/82353888-54b44800-9a00-11ea-9d5d-6c0d7c7ee387.jpeg)

So I managed to implement a simple renderer. It works, but there is still some more work to be done before I can ship it within the app -

- [x] Allow Multiple KatexWidgets to be used at once.
- [x] Cache the generated images to avoid rendering (it's very expensive)
- [ ] Integrate this into the Flutter Markdown renderer (will require forking the renderer)
- [x] Render the image with a transparent background. Currently it's rendered with a white background.
- [ ] Complete offline support - Currently it downloads an html file with the katex scripts each time.

**Comment by ModProg** [comment]
I second that, but I would suppose AsciiMath support as well, while not as powefull as Latex, I like it better in Markdown Notes due to it's readability.

This could maybe be achieved by converting to Latex:
[asciimath-to-latex](https://github.com/tylerlong/asciimath-to-latex) (MIT license in package.json probably should be translated to dart)

**Comment by c-antin** [comment]
Those 2 flutter packages seem promising:
https://pub.dev/packages/katex_flutter
https://pub.dev/packages/flutter_math_fork

**Comment by MalcolmMielle** [comment]
Was wondering where this issue was at? Is there a proof-of-concept somewhere? Would integrating one of the two flutter package showcased by @c-antin work?

**Comment by vHanda** [comment]
> Was wondering where this issue was at? Is there a proof-of-concept somewhere? Would integrating one of the two flutter package showcased by @c-antin work?

I've just spent about 90 minutes looking into this. The main blocker is getting the markdown parser to understand katex. I managed for the inline syntax, but not for the block one. I'll see if I can spend some more time tomorrow to make, at least, the inline version shippable.

The plus side of figuring this out is that it opens the door for further compatibility with Obsidian's markdown format, as I'll better understand how to modify the markdown parser.

**Comment by MalcolmMielle** [comment]
Sounds good, I was just curious :). Thanks for the great app!

**Comment by vHanda** [comment]
I've added Inline Katex Support. You can try out the dev version via FDroid - https://gitjournal.io/fdroid 

I still have to figure out the block syntax, also I'm leaning towards making this a pro feature.

**Comment by MalcolmMielle** [comment]
That sounds fair, it's a very specific feature :)

**Comment by MalcolmMielle** [comment]
Thanks for the inline parser, i tried it and it works really well it's very useful! 

However, not having the block syntax is a little awkward because it renders all annotations in the text but formulas don't appear. The note end up with situation like:

With pi, gamma, delta (<- represented correctly):

Massive formula (<- Raw latex)

I see from the commit comment that it's hard to implement though :S.
